### PR TITLE
debug console

### PR DESCRIPTION
**Pull request by akosyakov**
console

TODO:
- [x] input history
- [x] ANSI-output pretty-rendering
- [x] render variables tree
- [x] variables pretty-rendering
- [x] expressions pretty-rendering
- [x] theming
- [x] ~figure out whether it's worth to have more than one~ - decided to go with a simple approach with one console and adjust it when real use cases come up
- [x] multiline input handling
- [x] context menu
  - [x] clear
  - [x] copy
  - [x] ~copy all~ select all
  - [x] collapse all
- [x] evaluate expressions with a proper stack frame
- [x] monospace font
- [x] keyboard navigation to traverse variables
- [x] tree resizing issues
- [x] input context menu should not be trimmed
- [x] if any content widget, as suggest, is opened then on enter execute should not be available
- [x] widget title styling
- [x] completion support
- [x] run in terminals support
- [x] telemetry logging

Follow-up TODO:
- [ ] output source rendering & navigation
- [ ] links detection
- [ ] editor: evaluate selection
- [ ] indentation based output grouping

### CODE REVIEW

**Comment by tolusha** [comment]
We don't need this anymore:
https://github.com/theia-ide/theia/blob/master/packages/debug/src/browser/debug-session.ts#L304-L307

**Comment by tolusha** [comment]
@akosyakov 
Do we need a prefix to identify the current debug session ?
`Debug (attach) > some output`

**Comment by akosyakov** [comment]
@tolusha for now, I'm trying to align it with VS Code, they don't do it.

I am not sure that it is a good idea, because it will eat up output space and irrelevant in cases with a single running configuration. There can be other means to filter output by a running configuraiton, i.e. via a drop-down or checkboxes in the context menu.

// cc @svenefftinge

**Comment by akosyakov** [comment]
@thegecko added you to reviewers, I am still massaging code but you can try it out already

**Comment by akosyakov** [comment]
It's ready for the review. I will open an issue for follow-up enhancements.

**Comment by svenefftinge** [comment]
The fonts are two large. They should have the size of the editor font:
screen shot 2018-10-05 at 09 34 30

**Comment by svenefftinge** [comment]
The activation behaves odd. See gif below:

console-activation

**Comment by akosyakov** [comment]
> The activation behaves odd. See gif below:

@svenefftinge it focuses the input, what do you expect?

**Comment by svenefftinge** [comment]
> @svenefftinge it focuses the input, what do you expect?

It should also activate the tab.

**Comment by akosyakov** [comment]
@svenefftinge good catch! I've overridden `p-widget` class by mistake and it broke hiding.

I've addressed your comments, please have a look again.

File: packages/debug/src/browser/debug-session.ts [hunk 1]
```diff
@@ -175,9 +183,15 @@ export class DebugSessionImpl extends EventEmitter implements DebugSession {
...
 
+    completions(args: DebugProtocol.CompletionsArguments): Promise<DebugProtocol.CompletionsResponse> {
+        return this.proceedRequest('completions', args);
+    }
+
     protected handleMessage(data: string) {
         const message: DebugProtocol.ProtocolMessage = JSON.parse(data);
-        if (message.type === 'response') {
+        if (message.type === 'request') {
+            this.dispatchRequest(message as DebugProtocol.Request);
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/debug/src/browser/debug-session.ts]: me too, just picked `dispatch` for a minimal change, but renaming to `send` and `handle` sounds better

[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/debug/src/browser/debug-session.ts]: pushed the fix

**Pull request merged**